doc/subsystems/shell/shell.rst

```diff
@@ -0,0 +1,294 @@
+features:
+
+* Support for multiple instances.
+* Advanced cooperation with the :ref:`logger`.
+* Support for static and dynamic commands.
+* Smart command completion with the Tab key.
+* Built-in commands: ``clear``, ``shell``, ``colors``, ``echo``, ``history`` and ``resize``.
+* Viewing recently executed commands using the Up/Down keys.
+* Text edition using Left/Right/End/Home/Insert keys.
+* Support for ANSI escape codes: ``VT100`` and ``ESC[n~`` for cursor control and color printing.
+* Support for multiline commands.
+* Built-in handler to display help for the commands.
+* Support for wildcards: ``*`` and ``?``.
+* Support for meta keys.
+
+The module can be connected to any transport. At this point, the following
+transport layers are implemented:
+
+* UART
+
+For API documentation please refer :file:`include/shell/shell.h`
+
+Usage
+*****
+
+Use the :c:macro:`SHELL_DEFINE` macro to create an instance of the shell.
+It is important to pass correct newline character to this macro, otherwise
+the shell will not respond correctly to the Enter button.
+
+The following code shows a simple use case of this library:
+
+.. code-block:: c
+
+	/* Defining shell backend */
+	SHELL_UART_DEFINE(shell_transport_uart);
+
+	/* Creating shell instance */
+	SHELL_DEFINE(uart_shell, "uart:~$ ", &shell_transport_uart, '\r', 10);
+
+	void main(void)
+	{
+		(void)shell_init(&uart_shell, NULL, true, true, LOG_LEVEL_INF);
+	}
+
+	/* Creating subcommands (level 1 command) array for command "demo".
+	 * Subcommands must be added in alphabetical order
+	 */
+	SHELL_CREATE_STATIC_SUBCMD_SET(sub_demo)
+	{
+		/* Alphabetically sorted. */
+		SHELL_CMD(params, NULL, "Print params command.",
+						       cmd_demo_params),
+		SHELL_CMD(ping,   NULL, "Ping command.", cmd_demo_ping),
+		SHELL_SUBCMD_SET_END /* Array terminated. */
+	};
+	/* Creating root (level 0) command "demo" */
+	SHELL_CMD_REGISTER(demo, &sub_demo, "Demo commands", NULL);
+
+	/* Creating root (level 0) command "version" */
+	SHELL_CMD_REGISTER(version, NULL, "Show kernel version", cmd_version);
+
+
+Users may use the Tab key to complete a command/subcommand or to see the
+available subcommands for the currently entered command level.
+For example, when the cursor is positioned at the beginning of the command
+line and the Tab key is pressed, the user will see all root (level 0)
+commands:
+
+.. code-block:: none
+
+	  clear  demo  shell  history  log  resize  version
+
+
+.. note::
+	To view the subcommands that are available for a specific command, you must first type a space after this command and then hit Tab.
+
+These commands are registered by various modules, for example:
+
+* ``clear``, ``shell``, ``history``, and ``resize`` are built-in commands which have been registered by :file:`subsys/shell/.shell.c` ,
+* ``demo`` and ``version`` have been registered in example code above by main.c,
+* log has been registered by :file:`subsys/logging/log_cmds.c`
```

Comment:
to be consistent, add the double backtics:

    * ``log`` has been registered ...